bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java

```diff
@@ -190,11 +321,25 @@ public void store(final Item item, @Nullable final String alias) {
         }
 
         long now = System.currentTimeMillis() / 1000;
+        storageMap.computeIfAbsent(now, t -> new ConcurrentHashMap<>()).put(name, value);
+    }
 
-        scheduler.schedule(() -> internalStore(name, value, now, true), 0, TimeUnit.SECONDS);
+    private void doStore(boolean force) {
+        while (!storageMap.isEmpty()) {
+            long timestamp = storageMap.firstKey();
+            long now = System.currentTimeMillis() / 1000;
+            if (now > timestamp || force) {
+                // no new elements can be added for this timestamp because we are already past that time or the service
+                // requires forced storing
+                Map<String, Double> values = storageMap.pollFirstEntry().getValue();
+                values.forEach((name, value) -> writePointToDatabase(name, value, timestamp));
+            } else {
+                return;
```

Comment:
Could we run into trouble here once we have forecasts (or some other reason why we have a timestamp in the future) and this while loop runs in circles as the storageMap is not getting empty any time soon?

Reply:
Ah, I somehow didn't consider the `return` here. Yeah, sure.